File: server/agent/contracts.ts
import { z } from "zod";
import type { ImageAnalysis } from "../image-analysis-tool.js";

export type AnalysisHandoff = ImageAnalysis;
export type BodyProfile = AnalysisHandoff["body_profile"];
export type OutfitProfile = AnalysisHandoff["outfit_profile"];

export const sceneTypeSchema = z.enum(["mirror", "entrance_screen", "staff_ipad"]);
export const routeSchema = z.enum(["explicit", "recommendation", "unclear"]);
export const sessionStatusSchema = z.enum([
  "communicating",
  "recommending",
  "previewing",
  "confirmed",
  "handoff_ready",
  "staff_takeover",
  "ended"
]);
export const recommendationTypeSchema = z.enum(["explicit_need", "similar", "style", "seasonal"]);
export const feedbackTypeSchema = z.enum([
  "reject_all",
  "partial_adjust",
  "positive_keep",
  "confirm",
  // "换掉这件上衣，其余保留" — swap ONE slot, keep the rest. The slot to replace
  // rides in `dimension_value` (an OutfitSlotName).
  "swap_slot"
]);
export const feedbackDimensionSchema = z.enum(["color", "fit", "style", "price", "overall"]);
export const faceModeSchema = z.enum(["real_face", "default_face"]);
export const outfitSlotSchema = z.enum(["outerwear", "top", "bottom", "dress", "shoes", "accessory"]);

export type SceneType = z.infer<typeof sceneTypeSchema>;
export type Route = z.infer<typeof routeSchema>;
export type SessionStatus = z.infer<typeof sessionStatusSchema>;
export type RecommendationType = z.infer<typeof recommendationTypeSchema>;
export type FeedbackType = z.infer<typeof feedbackTypeSchema>;
export type FeedbackDimension = z.infer<typeof feedbackDimensionSchema>;
export type FaceMode = z.infer<typeof faceModeSchema>;
export type OutfitSlotName = z.infer<typeof outfitSlotSchema>;

export interface BodyTemplateResult {
  template_id: string;
  confidence: number;
  source: "match_body_template";
  notes: string;
}

export interface AgentConstraint {
  dimension: FeedbackDimension | "category" | "body_template" | "occasion" | "formality" | "color_tone";
  value: string;
  reason: string;
}

export interface AgentConstraints {
  prefer: AgentConstraint[];
  avoid: AgentConstraint[];
  budget_yen?: number;
}

export interface ParsedNeed {
  raw_text: string;
  categories: string[];
  colors: string[];
  style_tags: string[];
  occasion?: string;
  budget_yen?: number;
  // "正式/休闲" → formality; "深色/浅色" → color tone. Coarse facets the customer
  // asks for that map onto the derived product facets.
  formality?: string;
  color_tone?: string;
}

export interface ConstraintDelta {
  prefer: AgentConstraint[];
  avoid: AgentConstraint[];
  budget_yen?: number;
  requires_new_recommendation: boolean;
  notes: string;
}

export interface Product {
  product_id: string;
  sku: string;
  name: string;
  category: OutfitSlotName;
  audience?: string;
  price_yen: number;
  colors: string[];
  style_tags: string[];
  // Derived facets for coarse matching: formality (casual | smart_casual |
  // formal) and colour tone (dark | light | mixed). Computed from style_tags /
  // colors so the agent can honour "正式一点" / "深色系" requests directly.
  formality?: string;
  color_tone?: string;
  body_template_tags: string[];
  seasonal_rank: number;
  stock: Record<string, number>;
  image_url: string;
  vton_reference_image_url: string;
  vton_prompt: string;
}

export interface OutfitSlot {
  slot: OutfitSlotName;
  product_id: string;
  image_url: string;
  vton_reference_image_url: string;
  prompt: string;
}

export interface OutfitPayload {
  slots: OutfitSlot[];
}

export interface RecommendationSet {
  set_id: string;
  round: number;
  rec_type: RecommendationType;
  products: Product[];
  outfit: OutfitPayload;
  reason: string;
}

export interface RecommendationResponse {
  session_id: string;
  round: number;
  route: Route;
  sets: RecommendationSet[];
  warnings: string[];
}

export interface FaceConsentState {
  consent_given: boolean;
  face_mode: FaceMode;
  face_profile_id?: string;
  expire_at?: string;
}

export interface AhaDemoState {
  stage: "idle" | "lucy_preview" | "google_generating" | "google_ready" | "handoff_ready";
  lucy_preview_started_at?: string;
  lucy_preview_limit_sec: number;
  google_generation_status: "idle" | "queued" | "generating" | "ready";
  google_generation_eta_sec?: number;
  narrative: string;
}

export interface LucyRealtimeTryonPayload {
  provider: "decart_lucy_vton";
  mode: "realtime";
  configured: boolean;
  model: string;
  session_id: string;
  set_id: string;
  slot: OutfitSlotName;
  product_id: string;
  garment_image_url: string;
  prompt: string;
  duration_limit_sec: number;
  enhance: boolean;
  client_token: string | null;
  expires_at: string | null;
  warnings: string[];
}

export interface GoogleTryonFallbackPayload {
  provider: "google_vertex_tryon";
  mode: "static_or_async_video";
  session_id: string;
  set_id: string;
  template_id: string;
  outfit: OutfitPayload;
}

export interface TryonHandoffPayload {
  session_id: string;
  set_id: string;
  template_id: string;
  outfit: OutfitPayload;
  use_own_face: boolean;
  user_face: { face_profile_id: string } | null;
  lucy_realtime_preview: LucyRealtimeTryonPayload | null;
  google_fallback: GoogleTryonFallbackPayload;
}

export interface ToolCallRecord {
  tool: string;
  input: Record<string, unknown>;
  output: Record<string, unknown>;
  called_at: string;
}

export const startAgentSessionSchema = z.object({
  scene_type: sceneTypeSchema.default("mirror"),
  store_id: z.string().min(1).default("store_001"),
  analysis: z.custom<AnalysisHandoff>().optional()
});

export const agentChatSchema = z.object({
  text: z.string().min(1)
});

export const previewTryonSchema = z.object({
  set_id: z.string().min(1),
  slot: outfitSlotSchema.optional(),
  duration_limit_sec: z.number().int().min(3).max(30).default(12)
});

export const feedbackSourceSchema = z.enum([
  "voice",
  "quick_tag",
  "ui_click",
  "staff_input",
  "system"
]);
export type FeedbackSource = z.infer<typeof feedbackSourceSchema>;

export const feedbackPayloadSchema = z.object({
  set_id: z.string().min(1),
  feedback_type: feedbackTypeSchema.default("partial_adjust"),
  dimension: feedbackDimensionSchema.optional(),
  dimension_value: z.string().optional(),
  raw_voice_text: z.string().optional(),
  // Shared TEAM_CONTRACTS `source` enum; optional here since the local runtime
  // does not branch on it, but accepted so callers can pass it through.
  source: feedbackSourceSchema.optional()
});

export const confirmPayloadSchema = z.object({
  set_id: z.string().min(1),
  camera_processing_consent: z.boolean().default(false),
  face_profile_consent: z.boolean().default(false),
  image_ref: z.string().optional()
});

export const previewStatusPayloadSchema = z.object({
  status: z.enum(["previewing", "stopped", "failed"]),
  reason: z.string().optional(),
  lucy_session_id: z.string().optional()
});

export const purchasePayloadSchema = z.object({
  set_id: z.string().optional()
});

File: agent_foundation/contracts.py
"""Shared contract models for the Agent foundation.

These dataclasses mirror docs/TEAM_CONTRACTS.md and docs/TOOL_SCHEMAS.md.
They intentionally use only the Python standard library for the first prototype.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any


class SceneType(StrEnum):
    MIRROR = "mirror"
    STOREFRONT_SCREEN = "storefront_screen"
    STAFF_IPAD = "staff_ipad"


class Route(StrEnum):
    EXPLICIT = "explicit"
    RECOMMENDATION = "recommendation"
    UNCLEAR = "unclear"


class SessionStatus(StrEnum):
    ANALYZING = "analyzing"
    COMMUNICATING = "communicating"
    RECOMMENDING = "recommending"
    REFINING = "refining"
    CONFIRMED = "confirmed"
    ENDED = "ended"


class RecommendationType(StrEnum):
    SIMILAR = "similar"
    STYLE = "style"
    SEASONAL = "seasonal"
    EXPLICIT_NEED = "explicit_need"


class FeedbackType(StrEnum):
    REJECT_ALL = "reject_all"
    PARTIAL_ADJUST = "partial_adjust"
    POSITIVE_KEEP = "positive_keep"
    CONFIRM = "confirm"


class FeedbackDimension(StrEnum):
    COLOR = "color"
    FIT = "fit"
    STYLE = "style"
    PRICE = "price"
    OVERALL = "overall"


class FaceMode(StrEnum):
    REAL_FACE = "real_face"
    DEFAULT_FACE = "default_face"


class ToolStatus(StrEnum):
    SUCCESS = "success"
    PARTIAL = "partial"
    FAILED = "failed"


class InputSource(StrEnum):
    VOICE = "voice"
    QUICK_TAG = "quick_tag"
    UI_CLICK = "ui_click"
    STAFF_INPUT = "staff_input"
    SYSTEM = "system"


@dataclass(slots=True)
class BudgetRange:
    min: int = 0
    max: int | None = None
    currency: str = "JPY"

    def to_dict(self) -> dict[str, Any]:
        return {"min": self.min, "max": self.max, "currency": self.currency}


@dataclass(slots=True)
class ConstraintItem:
    dimension: str
    value: str
    reason: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "dimension": self.dimension,
            "value": self.value,
            "reason": self.reason,
        }


@dataclass(slots=True)
class AnalysisResult:
    matched_body_template_id: str
    current_style: list[str]
    confidence: float
    dominant_colors: list[str] = field(default_factory=list)
    analysis_id: str | None = None
    photo_url: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "matched_body_template_id": self.matched_body_template_id,
            "current_style": list(self.current_style),
            "confidence": self.confidence,
        }
        if self.dominant_colors:
            data["dominant_colors"] = list(self.dominant_colors)
        if self.analysis_id:
            data["analysis_id"] = self.analysis_id
        if self.photo_url:
            data["photo_url"] = self.photo_url
        return data


@dataclass(slots=True)
class UserNeedConstraints:
    category: list[str] = field(default_factory=list)
    style_tags: list[str] = field(default_factory=list)
    colors: list[str] = field(default_factory=list)
    budget_range: BudgetRange | None = None
    occasion: str | None = None
    avoid: list[ConstraintItem] = field(default_factory=list)
    keep: list[ConstraintItem] = field(default_factory=list)
    free_text_summary: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "category": list(self.category),
            "style_tags": list(self.style_tags),
            "colors": list(self.colors),
            "budget_range": self.budget_range.to_dict() if self.budget_range else None,
            "occasion": self.occasion,
            "avoid": [item.to_dict() for item in self.avoid],
            "keep": [item.to_dict() for item in self.keep],
            "free_text_summary": self.free_text_summary,
        }


@dataclass(slots=True)
class SessionContext:
    session_id: str
    scene_type: SceneType
    store_id: str
    round: int
    idempotency_key: str
    route: Route | None = None
    status: SessionStatus | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "session_id": self.session_id,
            "scene_type": self.scene_type.value,
            "store_id": self.store_id,
            "round": self.round,
            "idempotency_key": self.idempotency_key,
        }
        if self.route:
            data["route"] = self.route.value
        if self.status:
            data["status"] = self.status.value
        return data


@dataclass(slots=True)
class ProductSummary:
    product_id: str
    name: str
    category: str
    price: int
    currency: str
    color: str
    image_url: str
    in_stock: bool
    sku: str | None = None
    sizes: list[str] = field(default_factory=list)
    style_tags: list[str] = field(default_factory=list)
    body_fit_tags: list[str] = field(default_factory=list)
    fabric: str | None = None
    store_location: str | None = None
    stock_quantity: int | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "product_id": self.product_id,
            "name": self.name,
            "category": self.category,
            "price": self.price,
            "currency": self.currency,
            "color": self.color,
            "image_url": self.image_url,
            "in_stock": self.in_stock,
        }
        optional = {
            "sku": self.sku,
            "sizes": self.sizes,
            "style_tags": self.style_tags,
            "body_fit_tags": self.body_fit_tags,
            "fabric": self.fabric,
            "store_location": self.store_location,
            "stock_quantity": self.stock_quantity,
        }
        for key, value in optional.items():
            if value not in (None, [], {}):
                data[key] = value
        return data


@dataclass(slots=True)
class RecommendationSet:
    set_id: str
    rec_type: RecommendationType
    products: list[ProductSummary]
    reason: str
    title: str | None = None
    total_price: int | None = None
    currency: str = "JPY"
    score: float | None = None
    warnings: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        total_price = self.total_price
        if total_price is None:
            total_price = sum(product.price for product in self.products)
        data: dict[str, Any] = {
            "set_id": self.set_id,
            "rec_type": self.rec_type.value,
            "products": [product.to_dict() for product in self.products],
            "reason": self.reason,
            "total_price": total_price,
            "currency": self.currency,
        }
        if self.title:
            data["title"] = self.title
        if self.score is not None:
            data["score"] = self.score
        if self.warnings:
            data["warnings"] = list(self.warnings)
        return data


@dataclass(slots=True)
class RecommendationResponse:
    status: ToolStatus
    session_id: str
    round: int
    sets: list[RecommendationSet]
    applied_constraints: dict[str, Any] = field(default_factory=dict)
    warnings: list[str] = field(default_factory=list)
    debug_trace_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "status": self.status.value,
            "session_id": self.session_id,
            "round": self.round,
            "sets": [item.to_dict() for item in self.sets],
            "applied_constraints": dict(self.applied_constraints),
            "warnings": list(self.warnings),
        }
        if self.debug_trace_id:
            data["debug_trace_id"] = self.debug_trace_id
        return data


@dataclass(slots=True)
class FeedbackPayload:
    session_id: str
    set_id: str
    feedback_type: FeedbackType
    source: InputSource
    dimension: FeedbackDimension | None = None
    dimension_value: str | None = None
    raw_voice_text: str | None = None
    target_product_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "session_id": self.session_id,
            "set_id": self.set_id,
            "feedback_type": self.feedback_type.value,
            "source": self.source.value,
        }
        if self.dimension:
            data["dimension"] = self.dimension.value
        if self.dimension_value:
            data["dimension_value"] = self.dimension_value
        if self.raw_voice_text:
            data["raw_voice_text"] = self.raw_voice_text
        if self.target_product_id:
            data["target_product_id"] = self.target_product_id
        return data


@dataclass(slots=True)
class ConstraintDelta:
    avoid_add: list[ConstraintItem] = field(default_factory=list)
    keep_add: list[ConstraintItem] = field(default_factory=list)
    requires_new_recommendation: bool = True
    budget_update: BudgetRange | None = None
    style_shift: str | None = None
    fit_update: str | None = None
    notes: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "avoid_add": [item.to_dict() for item in self.avoid_add],
            "keep_add": [item.to_dict() for item in self.keep_add],
            "requires_new_recommendation": self.requires_new_recommendation,
            "budget_update": self.budget_update.to_dict() if self.budget_update else None,
            "style_shift": self.style_shift,
            "fit_update": self.fit_update,
            "notes": self.notes,
        }


@dataclass(slots=True)
class FaceConsentState:
    consent_given: bool
    face_mode: FaceMode
    face_profile_id: str | None = None
    default_face_template_id: str | None = None
    expire_at: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "consent_given": self.consent_given,
            "face_mode": self.face_mode.value,
            "face_profile_id": self.face_profile_id,
            "default_face_template_id": self.default_face_template_id,
            "expire_at": self.expire_at,
        }
